Slicing columns with mixed types <str>,<int> fails with ValueError

#### Code Sample, a copy-pastable example if possible

```python
In []:
df = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
df.loc[:, 'test':]
Out[]:
   test  1  2
0     1  2  3

In []:
df.loc[:, 1:]
Out[]:
TypeError                                 Traceback (most recent call last)
<ipython-input-655-3b97c047dcab> in <module>()
      1 df = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
----> 2 df.loc[:, 1:]
[snip]
/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/pandas/core/indexes/base.py in _invalid_indexer(self, form, key)
   1574                         "indexers [{key}] of {kind}".format(
   1575                             form=form, klass=type(self), key=key,
-> 1576                             kind=type(key)))
   1577 
   1578     def get_duplicates(self):

TypeError: cannot do slice indexing on <class 'pandas.core.indexes.base.Index'> with these indexers [1] of <class 'int'>

In []:
df = pd.DataFrame({0:1, 1:2, 2:3}, index=[0])
df.loc[:, 1:]
Out[]:
   1  2
0  2  3
```
#### Problem description

[this should explain **why** the current behaviour is a problem and why the expected output is a better solution.]
When having a columns with mixed types, e.g. `<str>, <int>` any slicing using the `int` columns causes a `ValueError`.

When you have just `int` columns slicing by `int` does not cause an issue.

**Note**: We receive a lot of issues on our GitHub tracker, so it is very possible that your issue has been posted before. Please check first before submitting so that we do not have to handle and close duplicates!

**Note**: Many problems can be resolved by simply upgrading `pandas` to the latest version. Before submitting, please check if that solution works for you. If possible, you may want to check if `master` addresses this issue, but that is not necessary.

For documentation-related issues, you can check the latest versions of the docs on `master` here:

https://pandas-docs.github.io/pandas-docs-travis/

If the issue has not been resolved there, go ahead and file it in the issue tracker.

#### Expected Output

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]
INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 17.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.22.0
pytest: 3.2.2
pip: 10.0.1
setuptools: 39.0.1
Cython: 0.28.1
numpy: 1.14.2
scipy: 1.0.0
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.7.2
patsy: None
dateutil: 2.6.1
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: 2.2.2
openpyxl: None
xlrd: None
xlwt: 1.3.0
xlsxwriter: None
lxml: None
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: 0.6.0
</details>

I tried adding a test under `class TestLabelSlicing` as below:
```
def test_loc_getitem_slice_columns_mixed_dtype(self):
    df1 = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
    df2 = pd.DataFrame({1:2, 2:3}, index=[0])
    tm.assert_equal(df1.loc[:, 1:], df2)
````

and got this error:
```
AssertionError: DataFrame.columns are different

DataFrame.columns classes are not equivalent
[left]:  Index([1, 2], dtype='object')
[right]: Int64Index([1, 2], dtype='int64')
```

I understand pandas creates different indices when there's mixed dtype. The question is should I change from `tm.assert_equal` to other assertion methods, like `assert df1.loc[:, 1:].equals(df2)`? Or should I stay with `tm.assert_equal` and try to work around the dtype difference?

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1645,10 +1645,10 @@ class TestLabelSlicing:
         tm.assert_equal(result, expected)
 
     def test_loc_getitem_slice_columns_mixed_dtype(self):
-        df1 = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
-        df2 = pd.DataFrame({1:2, 2:3}, index=[0])
-        breakpoint()
-        tm.assert_equal(df1.loc[:, 1:], df2)
+        # GH: 20975
+        df = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
+        expected = pd.DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2], dtype=object))
+        tm.assert_equal(df.loc[:, 1:], expected)
 
 
 class TestLocBooleanMask:

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1646,8 +1646,8 @@ class TestLabelSlicing:
 
     def test_loc_getitem_slice_columns_mixed_dtype(self):
         # GH: 20975
-        df = pd.DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
-        expected = pd.DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2],
+        df = DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
+        expected = DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2],
                                 dtype=object))
         tm.assert_equal(df.loc[:, 1:], expected)
 

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1646,9 +1646,10 @@ class TestLabelSlicing:
 
     def test_loc_getitem_slice_columns_mixed_dtype(self):
         # GH: 20975
-        df = DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
-        expected = DataFrame(data=[[2, 3]], index=[0], columns=pd.Index([1, 2],
-                                dtype=object))
+        df = DataFrame({"test": 1, 1: 2, 2: 3}, index=[0])
+        expected = DataFrame(
+            data=[[2, 3]], index=[0], columns=pd.Index([1, 2], dtype=object)
+        )
         tm.assert_equal(df.loc[:, 1:], expected)
 
 

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1647,7 +1647,7 @@ class TestLabelSlicing:
     def test_loc_getitem_slice_columns_mixed_dtype(self):
         # GH: 20975
         df = DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
-        expected = DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2],
+        expected = DataFrame(data=[[2, 3]], index=[0], columns=pd.Index([1, 2],
                                 dtype=object))
         tm.assert_equal(df.loc[:, 1:], expected)
 

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1650,7 +1650,7 @@ class TestLabelSlicing:
         expected = DataFrame(
             data=[[2, 3]], index=[0], columns=pd.Index([1, 2], dtype=object)
         )
-        tm.assert_equal(df.loc[:, 1:], expected)
+        tm.assert_frame_equal(df.loc[:, 1:], expected)
 
 
 class TestLocBooleanMask:

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1644,6 +1644,12 @@ class TestLabelSlicing:
         expected = frame_or_series(range(4), index=[value, "first", 2, "third"])
         tm.assert_equal(result, expected)
 
+    def test_loc_getitem_slice_columns_mixed_dtype(self):
+        df1 = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
+        df2 = pd.DataFrame({1:2, 2:3}, index=[0])
+        breakpoint()
+        tm.assert_equal(df1.loc[:, 1:], df2)
+
 
 class TestLocBooleanMask:
     def test_loc_setitem_bool_mask_timedeltaindex(self):

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1648,7 +1648,7 @@ class TestLabelSlicing:
         # GH: 20975
         df = pd.DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
         expected = pd.DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2],
-            dtype=object))
+                                dtype=object))
         tm.assert_equal(df.loc[:, 1:], expected)
 
 

--- a/pandas/tests/indexing/test_loc.py
+++ b/pandas/tests/indexing/test_loc.py
@@ -1646,8 +1646,9 @@ class TestLabelSlicing:
 
     def test_loc_getitem_slice_columns_mixed_dtype(self):
         # GH: 20975
-        df = pd.DataFrame({'test':1, 1:2, 2:3}, index=[0])
-        expected = pd.DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2], dtype=object))
+        df = pd.DataFrame({'test': 1, 1: 2, 2: 3}, index=[0])
+        expected = pd.DataFrame(data=[[2, 3]], index=[0], columns=Index([1, 2],
+            dtype=object))
         tm.assert_equal(df.loc[:, 1:], expected)
 
 

